common/transport/amqp/src/receiver_link.ts

```diff
@@ -80,98 +149,236 @@ export class ReceiverLink  extends EventEmitter implements AmqpLink {
+          },
+          attach: (null),
+          detach: (callback, err) => {
+            /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_06_002: [If the `detach` method is invoked on the `ReceiverLink` while still attaching, the ReceiverLink shall detach.  It will indicate the error to the callback for the `detach` as well as the callback to the `attach`.] */
+            debug('Detaching while attaching of rhea receiver link ' + this._linkAddress);
+            manageReceiverHandlers('removeListener');
+            //
+            // We may have a callback outstanding on the request that started the attaching.
+            // We will signal to that callback that an error occurred. We will also invoke the callback supplied
+            // for this detach.
+            //
+            let error = err || this._indicatedError || new Error('Unexpected link detach while attaching');
+            let attachingCallback = this._attachingCallback;
+            this._indicatedError = undefined;
+            this._attachingCallback = undefined;
+            if (attachingCallback) {
+              attachingCallback(error);
             }
-
-            this._fsm.transition('detached', undefined, err);
+            this._fsm.transition('detached', callback, error);
+          },
+          forceDetach: (err) => {
+            /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_06_003: [If the `forceDetach` method is invoked on the `ReceiverLink` while still attaching, the ReceiverLink shall detach.  It will indicate the error to the callback to the `attach`.] */
+            debug('Force detaching while attaching of rhea receiver link ' + this._linkAddress);
+            manageReceiverHandlers('removeListener');
+            let error = err || this._indicatedError;
+            let attachingCallback = this._attachingCallback;
+            this._indicatedError = undefined;
+            this._attachingCallback = undefined;
+            attachingCallback(error);
+            this._fsm.transition('detached', undefined, error);
           },
           accept: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
           reject: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
           abandon: (message, callback) => callback(new errors.DeviceMessageLockLostError())
         },
         attached: {
           _onEnter: (callback, err) => {
-            this._safeCallback(callback, err);
+            if (callback) callback(err);
+          },
+          receiverErrorEvent: (context: EventContext) => {
+            debug('in receiver attached state - error event for ' + context.receiver.name + ' error is: ' + this._getErrorName(context.receiver.error));
+            this._indicatedError = context.receiver.error;
+            //
+            // We don't transition at this point in that we are guaranteed that the error will be followed by a receiver_close
+            // event.
+            //
```

Comment:
does that need an intermediate state that is "errored but not closed yet"?

Reply:
as the comment above says we are guaranteed that close will follow.  It's not at all clear to me how error information is propagated around.  Even so, to introduce another state would double code in these sections.

---
In reply to: [209360561](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209360561) [](ancestors = 209360561)